perun-core/src/test/java/cz/metacentrum/perun/core/entry/GroupsManagerEntryIntegrationTest.java

```diff
@@ -245,6 +245,18 @@ public void expireMemberInSubGroupAndInUpperGroup() throws Exception {
 				foundMember.getGroupStatuses().get(group.getId()), MemberGroupStatus.EXPIRED);
 	}
 
+	@Test (expected=GroupRelationNotAllowed.class)
+	public void createGroupUnionForIndirectRelationship() throws Exception {
+		System.out.println(CLASS_NAME + "createGroupUnionForIndirectRelationship");
+
+		Vo vo = setUpVo();
+		groupsManagerBl.createGroup(sess, vo, group2);
+		groupsManagerBl.createGroup(sess, group2, group3);
+		groupsManagerBl.createGroup(sess, group3, group4);
+
+		groupsManagerBl.createGroupUnion(sess, group4, group2, false);
```

Comment:
I think there is an invalid order of the groups. The way it is implemented now, this test creates this relation:
2
|
3
|
4
And calling `createGroupUnion(sess, group4, group2, false)` tries to import members from group2 to group4. This causes a cycle, and I am not sure if we can detect this in the current implementation. Also, the test passes because of the error which I have pointed out above.
```suggestion
		groupsManagerBl.createGroupUnion(sess, group2, group4, false);
```

Reply:
Right, groups are switched here.